doc/user_guide/psy_data.rst

```diff
@@ -0,0 +1,294 @@
+#. ``PreStart`` indicating that a new instrumented code region is started.
+#. ``PreDeclareVariable`` is called for each variable that is written either
+   before or after the instrumented region.
+#. ``PreEndDeclaration`` to indicate the end of the variable declarations.
+#. ``ProvideVariable`` for each variable to be written before the instrumented
+   region.
+#. ``PreEnd`` to declare that all variables before the instrumented region
+   have been written.
+#. Then the actual instrumented code region is used.
+#. When the instrumented region is executed, a call to ``PostStart``
+   is added to indicate that all further variables written will be
+   after the user code.
+#. For each variable to be written after the user code a call to
+   ``ProvideVariable`` is added.
+#. A call to ``PostEnd`` is added once all variables have been written.
+
+.. note::
+    Depending on the options provided to the PSyData transformation
+    some of the calls might not be created. For example, for a performance
+    profiling library no variables will be declared or written.
+
+The following sections will describe the API in detail.
+
+.. _psy_data_type:
+
+``PSyDataType``
++++++++++++++++
+The library using the PSyData API must provide a user-defined data type
+called ``PSyDataType``. It is up to the application how this variable is
+used. PSyclone will declare the variables to be static, meaning that they
+can be used to accumulate data from call to call.
+
+``PreStart(this, module_name, kernel_name, num_pre_vars, num_post_vars)``
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+This function is called first each time the instrumented region is
+executed. It takes 4 parameters (besides the implicit PSyDataType
+instance):
+
+``module_name``::
+  This is the name of the module in which the original Fortran source
+  code is contained. Together with ``kernel_name`` it can be used to
+  create a unique name for each instrumented region.
+
+``kernel_name``::
+  The name of the kernel that is being executed.
+
+``num_pre_vars``::
+  This is the number of variables that will be supplied using
+  ``WriteVar`` before the instrumented region is executed.
+
+``num_post_vars``::
+  This is the number of variables that will be supplied using
+  ``WriteVar`` after the instrumented region is executed.
+  The sum ``num_pre_vars+num_post_vars`` is the number of
+  variable declarations that will follow.
+
+Typically the static ``PSyDataType`` instance can be used to store
+module and kernel name if they are required later, or to allocate
+arrays to store variable data.
+
+``PreDeclareVariable(this, name, value)``
++++++++++++++++++++++++++++++++++++++++++
+This function is called for each variable that will be written
+before or after the user instrumented region. If a variable
+is written before and after, it will also be declared twice
+(it can be useful to provide a variable using a different
```

Comment:
That depends on the library. E.g. netcdf probably does not allow me to declare a variable more than once. If you imagine a simple text file it could accommodate the same variable more than once (with the semantics: first time input, second time output). 

If you think of a 'compute checksum of read-only-fields' function, then it will (likely) be essential that each variable is provided twice - once to compute the checksum before calling, second time to recompute and verify the checksum. Maybe we could add more semantics when we introduce the classes of PSyData that is being discussed (or better: that we have agreed on).